File: polygon/src/main/java/com/xattacker/android/view/polygon/PolygonView.java
package com.xattacker.android.view.polygon;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Align;
import android.graphics.Paint.Style;
import android.graphics.Path;
import android.graphics.PointF;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class PolygonView extends View
{
	private ArrayList<PolygonRegion> _regions;
	private PolygonRegion _clickedRegion;
	private RegionMark _clickedMark;
	private WeakReference<PolygonViewListener> _listener;
	
	private int _borderColor, _highlightColor, _titleColor, _highlightMarkColor;
	private float _titleFontSize;
	private Paint _paint;
	private Path _path;
	private int _baseWidth, _baseHeight;
	private float _ratioW, _ratioH;
	private boolean _fitToCenter;

	public PolygonView(Context context, AttributeSet attrs)
	{
		super(context, attrs);
		
		initView();
	}
	
	public PolygonView(Context context)
	{
		super(context);

		initView();
	}
	
	public void setListener(PolygonViewListener aListener)
	{
		_listener = new WeakReference<PolygonViewListener>(aListener);
	}

	public void loadMap(PolygonMap aMap)
	{
		if (aMap != null)
		{
			_borderColor = aMap.getBorderColor();
			_highlightColor = aMap.getHighlightColor();
			_highlightMarkColor = aMap.getHighlightMarkColor();
			_titleColor = aMap.getTitleColor();
			setBaseSize(aMap.getWidth(), aMap.getHeight());
			
		   _regions.clear();
		   
		   if (aMap.getRegions() != null && !aMap.getRegions().isEmpty())
		   {
		   		addRegions(aMap.getRegions());
		   }
		}
	}

	public void addRegion(PolygonRegion aRegion)
	{
		if (aRegion != null)
		{
			_regions.add(aRegion);
			requestLayout();
			invalidate();
		}
	}

	public void addRegion(PolygonRegion... aRegions)
	{
		if (aRegions != null && aRegions.length > 0)
		{
			addRegions(Arrays.asList(aRegions));
		}
	}
	
	public void addRegions(List<PolygonRegion> aRegions)
	{
		if (aRegions != null && !aRegions.isEmpty())
		{
			_regions.addAll(aRegions);
			requestLayout();
			invalidate();
		}
	}

	public void clearRegions()
	{
		_regions.clear();
		_ratioW = _ratioH = 1;
		invalidate();
	}
	
	public void setBorderColor(int aColor)
	{
		_borderColor = aColor;
	}
	
	public void setHighlightColor(int aColor)
	{
		_highlightColor = aColor;
	}
	
	public void setTitleColor(int aColor)
	{
		_titleColor = aColor;
	}
	
	public void setTitleFontSize(float aSize)
	{
		_titleFontSize = aSize;
	}
	
	public void setBaseSize(int aWidth, int aHeight)
	{
		_baseWidth = aWidth;
		_baseHeight = aHeight;
	}
	
	public void setFitToCenter(boolean aCenter)
	{
		_fitToCenter = aCenter;
	}
	
	@Override
	public boolean onTouchEvent(MotionEvent event)
	{
		if (!isEnabled())
		{
			return false;
		}

		
		if (_regions != null && !_regions.isEmpty())
		{
			PointF point = new PointF(event.getX(), event.getY());

			switch (event.getAction())
			{
				case MotionEvent.ACTION_DOWN:
				{
					for (PolygonRegion region : _regions)
					{
						if (region.hasMark())
						{
							boolean hit = false;
							
							for (RegionMark mark : region._marks)
						   {
								if (mark.isPointInMark(point))
								{
									_clickedMark = mark;
									_clickedMark._belongRegion = region;
									invalidate();
									hit = true;
									
									break;
								}
						   }
							
							if (hit)
							{
								break;
							}
						}
						else if (region.isPointInRegion(point))
						{
							_clickedRegion = region;
							invalidate();
							
							break;
						}
					}
				}
					break;
				
				case MotionEvent.ACTION_UP:
				{
					if (_listener != null && _listener.get() != null)
					{
						if (_clickedMark != null)
						{
							_listener.get().onRegionMarkClicked(_clickedMark, _clickedMark._belongRegion);
						}
						else if (_clickedRegion != null)
						{
							_listener.get().onRegionClicked(_clickedRegion);
						}
					}
					
					_clickedRegion = null;
					_clickedMark = null;
					invalidate();
				}
					break;
					
				case MotionEvent.ACTION_CANCEL:
					_clickedRegion = null;
					_clickedMark = null;
					invalidate();
					break;
			}
		}
		
		return true;
	}

	@Override
	protected void onLayout(boolean changed, int l, int t, int r, int b)
	{
		super.onLayout(changed, l, t, r, b);

		if (_baseWidth != 0 && _baseHeight != 0)
		{
			float ratio_w = (float)getWidth()  / (float)_baseWidth;
			float ratio_h = (float)getHeight()  / (float)_baseHeight;

			if (ratio_w != _ratioW || ratio_h != _ratioH)
			{
				_ratioW = ratio_w;
				_ratioH = ratio_h;

				float min_x = getWidth(), min_y = getHeight(), max_x = 0, max_y = 0;
				float ratio = _ratioW > _ratioH ? _ratioH : _ratioW;

				for (PolygonRegion region : _regions)
				{
					if (!region._points.isEmpty())
					{
						for (PointF point : region._points)
						{
							point.x *= ratio;
							point.y *= ratio;

							if (_fitToCenter)
							{
								if (min_x > point.x)
								{
									min_x = point.x;
								}
								else if (max_x < point.x)
								{
									max_x = point.x;
								}

								if (min_y > point.y)
								{
									min_y = point.y;
								}
								else if (max_y < point.y)
								{
									max_y = point.y;
								}
							}
						}

						PointF point = region._titleInfo._position;
						if (point.x >= 0 && point.y >= 0)
						{
							point.x *= ratio;
							point.y *= ratio;
						}
					}


					if (region.hasMark())
					{
						PointF point = null;

						for (RegionMark mark : region._marks)
						{
							point = mark._position;
							point.x *= ratio;
							point.y *= ratio;
						}
					}
				}


				if (_fitToCenter)
				{
					float offset_x = getWidth()/2 - ((max_x + min_x)/2);
					float offset_y = getHeight()/2 - ((max_y + min_y)/2);

					if (offset_x != 0 || offset_y != 0)
					{
						fitCenter(offset_x, offset_y);
					}
				}
			}
		}
	}

	@Override
	protected void onDraw(Canvas canvas)
	{
		if (_regions != null && !_regions.isEmpty())
		{
			for (PolygonRegion region : _regions)
			{
				if (!region._points.isEmpty())
				{
					// draw region area
					_paint.setStyle(Paint.Style.FILL);
					_paint.setColor(region.getRegionColor());
					drawPath(canvas, region._points);
	
					if (_clickedRegion != null && _clickedRegion == region)
					{
						_paint.setColor(_highlightColor);
						drawPath(canvas, region._points);
					}
					
					// draw region border
					_paint.setStyle(Paint.Style.STROKE);
					_paint.setColor(_borderColor);
					drawPath(canvas, region._points);

					if (region._titleInfo.title != null && region._titleInfo.title.length() > 0)
					{
						// draw region title
						PointF p = region._titleInfo._position;
						if (p.x < 0 && p.y < 0)
						{
							 p = region.getCentral();
						}

						_paint.setTextAlign(Align.CENTER);
						_paint.setColor(_titleColor);
						_paint.setTextSize(_titleFontSize);
						canvas.drawText(region._titleInfo.title, p.x, p.y, _paint);
					}
				}
				
				
				if (region.hasMark())
				{
					// draw region mark
					_paint.setStyle(Style.FILL);
					_paint.setTextAlign(Align.CENTER);
					_paint.setTextSize(_titleFontSize);
					
					for (RegionMark mark : region._marks)
				   {
						_paint.setColor(_clickedMark != null && _clickedMark == mark ? _highlightMarkColor : mark.getColor()._color);
						canvas.drawCircle(mark._position.x, mark._position.y, RegionMark.MARK_RADIUS, _paint);
						
						if (mark._title != null && mark._title.length() > 0)
						{
							canvas.drawText(mark._title, mark._position.x, mark._position.y + RegionMark.MARK_RADIUS*2, _paint);
						}
				   }
				}
			}
		}
	}

	private void fitCenter(float aOffsetX, float aOffsetY)
	{
		for (PolygonRegion region : _regions)
		{
			if (!region._points.isEmpty())
			{
				PointF point = null;

				for (int i = 0, size = region._points.size(); i < size; i++)
				{
					point = region._points.get(i);
					point.x += aOffsetX;
					point.y += aOffsetY;
				}

				point = region._titleInfo._position;
				if (point.x >= 0 && point.y >= 0)
				{
					point.x += aOffsetX;
					point.y += aOffsetY;
				}
			}


			if (region.hasMark())
			{
				PointF point = null;

				for (RegionMark mark : region._marks)
				{
					point = mark._position;
					point.x += aOffsetX;
					point.y += aOffsetY;
				}
			}
		}
	}
	
	private void drawPath(Canvas aCanvas, ArrayList<PointF> aPoints)
	{	
		_path.reset();
		
		PointF point = aPoints.get(0);
		_path.moveTo(point.x, point.y); // first point
		
		for (int i = 1, size = aPoints.size(); i < size; i++)
		{
			point = aPoints.get(i);
			_path.lineTo(point.x, point.y);
		}
		
		_path.close();
		
		aCanvas.drawPath(_path, _paint);
	}
	
	private void initView()
	{
		_regions = new ArrayList<PolygonRegion>();
		_clickedRegion = null;
		_clickedMark = null;
		
		_borderColor = Color.DKGRAY;
		_highlightColor = 0x77000000 & Color.LTGRAY;
		_highlightMarkColor = 0x77000000 & Color.DKGRAY;
		_titleColor = Color.BLACK;
		_titleFontSize = 20;
		_ratioW = _ratioH = 0;
		_baseWidth = _baseHeight = 0;
		_path = new Path();
		_fitToCenter = true;
		 
		_paint = new Paint();
		_paint.setAntiAlias(true);
		_paint.setStyle(Paint.Style.FILL);
		_paint.setStrokeWidth(2);
	}
}

File: polygon/src/main/java/com/xattacker/android/view/polygon/PolygonView.kt
package com.xattacker.android.view.polygon

import android.content.Context
import android.graphics.*
import android.graphics.Paint.Align
import android.graphics.Paint.Style
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import java.lang.ref.WeakReference
import java.util.*
import kotlin.math.max
import kotlin.math.min

class PolygonView : View
{
    var borderColor: Int = 0
    var highlightColor: Int = 0
    var titleColor: Int = 0
    var highlightMarkColor: Int = 0
    var titleFontSize: Float = 0f
    var fitToCenter: Boolean = true

    var listener: PolygonViewListener?
        get() = _listener?.get()
        set(listener)
        {
            if (listener != null)
            {
                _listener = WeakReference(listener)
            }
        }

    private var _regions: ArrayList<PolygonRegion>? = null
    private var _clickedRegion: PolygonRegion? = null
    private var _clickedMark: RegionMark? = null
    private var _listener: WeakReference<PolygonViewListener>? = null

    private lateinit var _paint: Paint
    private lateinit var _path: Path
    private var _baseWidth: Int = 0
    private var _baseHeight: Int = 0
    private var _ratioW: Float = 0f
    private var _ratioH: Float = 0f

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    {
        initView()
    }

    constructor(context: Context) : super(context)
    {
        initView()
    }

    fun loadMap(aMap: PolygonMap)
    {
        borderColor = aMap.borderColor
        highlightColor = aMap.highlightColor
        highlightMarkColor = aMap.highlightMarkColor
        titleColor = aMap.titleColor
        setBaseSize(aMap.width, aMap.height)

        _regions?.clear()

        if (!aMap.regions.isEmpty())
        {
            addRegions(aMap.regions)
        }
    }

    fun addRegion(aRegion: PolygonRegion)
    {
        _regions?.add(aRegion)
        requestLayout()
        invalidate()
    }

    fun addRegion(vararg aRegions: PolygonRegion)
    {
        if (aRegions.size > 0)
        {
            addRegions(Arrays.asList(*aRegions))
        }
    }

    fun addRegions(aRegions: List<PolygonRegion>)
    {
        if (!aRegions.isEmpty())
        {
            _regions?.addAll(aRegions)
            requestLayout()
            invalidate()
        }
    }

    fun clearRegions()
    {
        _regions?.clear()
        _ratioH = 1f
        _ratioW = _ratioH
        invalidate()
    }

    fun setBaseSize(width: Int, height: Int)
    {
        _baseWidth = width
        _baseHeight = height
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!isEnabled) {
            return false
        }


        val regions = _regions
        if (regions != null && !regions.isEmpty())
        {
            val point = PointF(event.x, event.y)

            when (event.action) {
                MotionEvent.ACTION_DOWN ->
                {
                    for (region in regions)
                    {
                        if (region.hasMark)
                        {
                            var hit = false

                            region.marks?.let {
                                for (mark in it)
                                {
                                    if (mark.isPointInMark(point))
                                    {
                                        _clickedMark = mark
                                        _clickedMark?._belongRegion = region
                                        invalidate()
                                        hit = true

                                        break
                                    }
                                }
                            }

                            if (hit)
                            {
                                break
                            }
                        }
                        else if (region.isPointInRegion(point))
                        {
                            _clickedRegion = region
                            invalidate()

                            break
                        }
                    }
                }

                MotionEvent.ACTION_UP ->
                {
                    if (_clickedMark != null)
                    {
                        _clickedMark?.let {
                            mark ->
                            mark._belongRegion?.let {
                                region ->
                                _listener?.get()?.onRegionMarkClicked(mark, region)
                            }
                        }
                    }
                    else if (_clickedRegion != null)
                    {
                        _clickedRegion?.let {
                            _listener?.get()?.onRegionClicked(it)
                        }
                    }

                    _clickedRegion = null
                    _clickedMark = null
                    invalidate()
                }

                MotionEvent.ACTION_CANCEL ->
                {
                    _clickedRegion = null
                    _clickedMark = null
                    invalidate()
                }
            }
        }

        return true
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int)
    {
        super.onLayout(changed, l, t, r, b)

        if (_baseWidth == 0 || _baseHeight == 0)
        {
            return
        }


        val ratio_w = width.toFloat() / _baseWidth.toFloat()
        val ratio_h = height.toFloat() / _baseHeight.toFloat()
        if (ratio_w != _ratioW || ratio_h != _ratioH)
        {
            _ratioW = ratio_w
            _ratioH = ratio_h

            var min_x = width.toFloat()
            var min_y = height.toFloat()
            var max_x = 0f
            var max_y = 0f
            val ratio = if (_ratioW > _ratioH) _ratioH else _ratioW

            _regions?.let {
                regions ->
                for (region in regions)
                {
                    region.points?.let {
                        points ->
                        for (point in points)
                        {
                            point.x *= ratio
                            point.y *= ratio

                            if (this.fitToCenter)
                            {
                                min_x = min(min_x, point.x)
                                max_x = max(max_x, point.x)

                                min_y = min(min_y, point.y)
                                max_y = max(max_y, point.y)
                            }
                        }

                        val point = region.titleInfo.position
                        if (point.x >= 0 && point.y >= 0)
                        {
                            point.x *= ratio
                            point.y *= ratio
                        }
                    }

                    region.marks?.let {
                        marks ->
                        for (mark in marks)
                        {
                            val point = mark.position
                            point.x *= ratio
                            point.y *= ratio
                        }
                    }
                }
            }

            if (this.fitToCenter)
            {
                val offset_x = width / 2 - (max_x + min_x) / 2
                val offset_y = height / 2 - (max_y + min_y) / 2
                if (offset_x != 0f || offset_y != 0f)
                {
                    fitCenter(offset_x, offset_y)
                }
            }
        }
    }

    override fun onDraw(canvas: Canvas)
    {
        val regions = _regions
        if (regions == null || regions.isEmpty())
        {
            return
        }


        for (region in regions)
        {
            val points = region.points
            if (points?.isEmpty() == false)
            {
                // draw region area
                _paint.style = Paint.Style.FILL
                _paint.color = region.regionColor
                drawPath(canvas, points)

                if (_clickedRegion == region)
                {
                    _paint.color = highlightColor
                    drawPath(canvas, points)
                }

                // draw region border
                _paint.style = Paint.Style.STROKE
                _paint.color = borderColor
                drawPath(canvas, points)

                if (region.titleInfo.title?.length ?: 0 > 0) {
                    // draw region title
                    var p = region.titleInfo.position
                    if (p.x < 0 && p.y < 0)
                    {
                        p = region.central
                    }

                    _paint.textAlign = Align.CENTER
                    _paint.color = titleColor
                    _paint.textSize = titleFontSize
                    canvas.drawText(region.titleInfo.title ?: "", p.x, p.y, _paint)
                }
            }


            region.marks?.let {
                marks ->
                // draw region mark
                _paint.style = Style.FILL
                _paint.textAlign = Align.CENTER
                _paint.textSize = titleFontSize

                for (mark in marks)
                {
                    _paint.color = if (_clickedMark != null && _clickedMark == mark) highlightMarkColor else mark.color?.color ?: Color.TRANSPARENT
                    canvas.drawCircle(mark.position.x, mark.position.y, RegionMark.MARK_RADIUS, _paint)

                    val title = mark.title
                    if ( title?.length ?: 0 > 0)
                    {
                        canvas.drawText(title ?: "", mark.position.x, mark.position.y + RegionMark.MARK_RADIUS * 2, _paint)
                    }
                }
            }
        }
    }

    private fun fitCenter(offsetX: Float, offsetY: Float)
    {
        _regions?.let {
            regions ->
            for (region in regions)
            {
                region.points?.let {
                    for (point in it)
                    {
                        point.x += offsetX
                        point.y += offsetY
                    }

                    val point = region.titleInfo.position
                    if (point.x >= 0 && point.y >= 0)
                    {
                        point.x += offsetX
                        point.y += offsetY
                    }
                }

                region.marks?.let {
                    marks ->
                    for (mark in marks)
                    {
                        val point = mark.position
                        point.x += offsetX
                        point.y += offsetY
                    }
                }
            }
        }
    }

    private fun drawPath(aCanvas: Canvas, aPoints: ArrayList<PointF>)
    {
        _path.reset()

        val point = aPoints[0]
        _path.moveTo(point.x, point.y) // first point

        for (p in aPoints)
        {
            _path.lineTo(p.x, p.y)
        }

        _path.close()

        aCanvas.drawPath(_path, _paint)
    }

    private fun initView()
    {
        _regions = ArrayList()
        _clickedRegion = null
        _clickedMark = null

        borderColor = Color.DKGRAY
        highlightColor = 0x77000000 and Color.LTGRAY
        highlightMarkColor = 0x77000000 and Color.DKGRAY
        titleColor = Color.BLACK
        titleFontSize = 20f
        _ratioH = 0f
        _ratioW = _ratioH
        _baseHeight = 0
        _baseWidth = _baseHeight
        _path = Path()
        fitToCenter = true

        _paint = Paint()
        _paint.isAntiAlias = true
        _paint.style = Paint.Style.FILL
        _paint.strokeWidth = 2f
    }
}
